our-paxos/paxos.py

```diff
@@ -106,44 +103,69 @@ def acceptor(config, id):
+        init_state = {"rnd": 0, "v-rnd": 0, "v-val": 0}
         # recv to a large buffer
         msg = r.recv(2**16)
-        # extract the instance and phase
+        # extract the instance id and phase
         loc = paxos_decode(msg)
-        instance = loc[0]
+        id = loc[0]
         phase = loc[1]
         
         # set the initial state if it's the first time we start
         # this paxos instance
-        if instance not in paxos_instances.keys():
-            paxos_instances[instance] = init_state
-
-        # let's change the current state
-        state = paxos_instances[instance]
-        if phase == 1:
-            # received phase 1A msg from proposer
-            if loc[2] > state['rnd']:
-                # update state and paxos_
+        if id not in paxos_instances.keys():
+            paxos_instances[id] = init_state
+
+        # in python changes to state (a dict) will be changed also in
+        # paxos_intances (a dict of dicts). Using to avoid repeating
+        # paxos_intances[instance]
+        state = paxos_instances[id]
+        
+        if phase == 1: # received phase 1A msg from proposer
+            if loc[2] > state['rnd']: # loc[2] = c-rnd
                 state['rnd'] = loc[2]
-                print("i: {} p: {}, state: {}".format(instance, phase, state))
+                print("i: {} p: {}, state: {}".format(id, phase, state))
+
                 # encode and send phase 1B to proposer
-                msg = paxos_encode([instance, phase, state['rnd'], state['v-rnd'], state['v-val']])
+                msg = paxos_encode([id, phase, state['rnd'], state['v-rnd'], state['v-val']])
                 s.sendto(msg, config['proposers'])
-        elif phase == 2:
-            # received phase 2A msg from proposer
-            # loc[2] = c-rnd, loc[3] = c-val
-            if loc[2] >= state['rnd']:
+
+                # start timeout on message 2A
+                thread = Thread(target=acceptor_timeout, args=(id))
+                thread.start()
+                thread.join()
+        
+        elif phase == 2: # received phase 2A msg from proposer
+            
+             # loc[2] = c-rnd, loc[3] = c-val
+             if loc[2] >= state['rnd']:
                 #state['rnd'] = loc[2] # not in the slides, but it makes sense (?)
                 state['v-rnd'] = loc[2]
                 state['v-val'] = loc[3]
-                print("i: {} p: {}, state: {}".format(instance, phase, state))
+                print("i: {} p: {}, state: {}".format(id, phase, state))
 
                 # encode and send phase 2B to proposer
-                msg = paxos_encode([instance, phase, state['v-rnd'], state['v-val']])
+                msg = paxos_encode([id, phase, state['v-rnd'], state['v-val']])
                 s.sendto(msg, config['learners'])
+        
+        elif phase == 4: # received "resend 2B" request from learner timeout
+            msg = paxos_encode([id, 2, state['v-rnd'], state['v-val']])
```

Comment:
Where do you get state['v-rnd']? What if you have few instances going simultanuosly?

Reply:
discussed this